### ISSUE TITLE
Invalid typing for __init__ in Generic classes

### ISSUE THREAD

**Issue by last-partizan** [microsoft/pyright#2909]
**Bug description**
False positive for "Argument for type .. could not be assigned.

**To Reproduce**

Consider following example, this is minimal version of [django-stubs/db/models/related.pyi:ForeignKey](https://github.com/sbdchd/django-types/blob/946c286cf0b0197f2253f1aeb403f8a8dc9139ff/django-stubs/db/models/fields/related.pyi#L267)

```python
# related_field_minimal.py
from typing import Any, Generic, Literal, Optional, TypeVar, overload

_M = TypeVar("_M")


class ForeignKey(Generic[_M]):
    @overload
    def __new__(
        cls,
        null: Literal[False] = ...,
        **kwargs: Any,
    ) -> ForeignKey[_M]: ...

    @overload
    def __new__(
        cls,
        null: Literal[True],
        **kwargs: Any,
    ) -> ForeignKey[Optional[_M]]: ...
    @overload
    def __init__(
        self: ForeignKey[_M],
        null: Literal[False] = ...,
    ) -> None: ...
    @overload
    def __init__(
        self: ForeignKey[Optional[_M]],
        null: Literal[True] = ...,
    ) -> None: ...
```

```python
# models.py
from related_field_minimal import ForeignKey

class Test:
    pass

class Model:
    f1 = ForeignKey(null=True)
    f2 = ForeignKey["Test"](null=True)
```

Running `pyright` over this file, produces following error:
```
>     f2 = ForeignKey["Test"](null=True)
error: Argument of type "Literal[True]" cannot be assigned to parameter "null" of type "Literal[False]" in function "__init__"
```

**Expected behavior**
It should produce no error. (And it works in `mypy`)

**VS Code extension or command-line**

```
> pyright --version
pyright 1.1.212
```

**Comment by erictraut** [comment]
I think the error is correct. The TypeVar `_M` is invariant in the `ForeignKey` class. That means `ForeignKey["Test"]` is not compatible with `ForeignKey[Optional[_M]]`, so there is no `__init__` method overload that matches. If you use `ForeignKey[Optional[Test]]`, it works without error. Also, if you change `_M` to be covariant, it works without error.

**Comment by last-partizan** [comment]
Oh, that is how it works.

I think i understand it now. Yes, looks like pyright is right.

Thanks!

But i have some other thoughts:

I was thinking that `__new__` overrides are working this way:

`ForeignKey[Test](null=True)` will generate `ForeignKey(Test | None)`, and from logical point of view this is what i want to happen.

But even if i remove them (`__new__` overrides), mypy still gets the correct type.

I've added `to` field, to demonstrate how pyright correctly works in this example.
Also, i've changed `_M` to be bound to `Optional[Model]`, which i forgot to do earlier.

```python
# related_field_minimal.py
from typing import Generic, Literal, Optional, Type, TypeVar, overload

class Model: ...

_M = TypeVar("_M", bound=Optional[Model])

class ForeignKey(Generic[_M]):

    @overload
    def __init__(
        self: ForeignKey[_M],
        to: Type[_M],
        null: Literal[False] = ...,
    ) -> None: ...
    @overload
    def __init__(
        self: ForeignKey[Optional[_M]],
        to: Type[_M],
        null: Literal[True] = ...,
    ) -> None: ...
```

```python
# models.py
from typing import Optional
from related_field_minimal import ForeignKey, Model


class Test(Model):
    pass


class BookModel(Model):
    required_field = ForeignKey(to=Test, null=False)
    optional_f1 = ForeignKey(to=Test, null=True)
    optional_f2 = ForeignKey[Test](to=Test, null=True)
    optional_f3 = ForeignKey[Optional[Test]](to=Test, null=True)

reveal_type(BookModel.required_field)
reveal_type(BookModel.optional_f1)
reveal_type(BookModel.optional_f2)
reveal_type(BookModel.optional_f3)
```

```
# pyright
  /home/serg/src/django-types/test-related/models.py:12:50 - error: Argument of type "Literal[True]" cannot be assigned to parameter "null" of type "Literal[False]" in function "__init__"
    "Literal[True]" cannot be assigned to type "Literal[False]" (reportGeneralTypeIssues)
  /home/serg/src/django-types/test-related/models.py:15:13 - information: Type of "BookModel.required_field" is "ForeignKey[Test]"
  /home/serg/src/django-types/test-related/models.py:16:13 - information: Type of "BookModel.optional_f1" is "ForeignKey[_M@ForeignKey |None]"
  /home/serg/src/django-types/test-related/models.py:17:13 - information: Type of "BookModel.optional_f2" is "ForeignKey[Test]"
  /home/serg/src/django-types/test-related/models.py:18:13 - information: Type of "BookModel.optional_f3" is "ForeignKey[Test | None]"
1 error, 0 warnings, 4 informations
```

```
# mypy
models.py:15: note: Revealed type is "related_field_minimal.ForeignKey[models.Test]"
models.py:16: note: Revealed type is "related_field_minimal.ForeignKey[Union[models.Test, None]]"
models.py:17: note: Revealed type is "related_field_minimal.ForeignKey[Union[models.Test, None]]"
models.py:18: note: Revealed type is "related_field_minimal.ForeignKey[Union[models.Test, None]]"
```

Now pyright correctly recognizes nullable `optional_f1`, but still incorrectly works with `optional_f2`

**Comment by erictraut** [comment]
You've bound `_M` to `Optional[Model]`, and `_M` is an invariant TypeVar. That means that any type argument to `ForeignKey` must be assignable to `Model | None`. Therefore, `ForeignKey[Test]` is not a valid specialization of the class because `Test` is not assignable to `Model | None`. That explains why `ForeignKey[Test](to=Test, null=True)` fails. The error message is a bit opaque, but that's because this involves an overload, and overload-related error messages are difficult to express clearly.

**Comment by last-partizan** [comment]
Also, i was incorrectly stating that:

> Now pyright correctly recognizes nullable optional_f1, but still incorrectly works with optional_f2

In reality, `optional_f1` does not produce type errors, but type is not what i expect it to be:

```
    optional_f1 = ForeignKey(to=Test, null=True)

### mypy : correct
# models.py:16: note: Revealed type is "related_field_minimal.ForeignKey[Union[models.Test, None]]"

### pyright : unexpected
# "BookModel.optional_f1" is "ForeignKey[_M@ForeignKey |None]"
```

Maybe in your investigation it would be useful to look at original classes which i stripped down to minimal state from this:  https://github.com/sbdchd/django-types/blob/946c286cf0b0197f2253f1aeb403f8a8dc9139ff/django-stubs/db/models/fields/related.pyi#L267 (but errors are the same, and i can reproduce them on both - django-stubs and this stripped down version)

**Comment by last-partizan** [comment]
Btw, removing overloads from `__init__` will fix this error, becouse `__new__` clearly defines that `null=True` will generate `ForeignKey[Optional[_M]]`.

And probably we're gonna do it this way: https://github.com/sbdchd/django-types/issues/87

**Comment by erictraut** [comment]
The reason you're not getting the results that you're hoping for is because you're using an undocumented (and unspecified) typing feature — namely you're providing a type annotation for the `self` parameter in the `__init__` method and expecting this to inform the evaluated type of the constructor. PEP 484 doesn't specify this behavior, nor do any other PEPs. It is something that mypy introduced. I was very uncomfortable adding support for this in pyright because it is not documented and not specified, and there are a bunch of edge cases that are not well thought out. My hand was forced when some of the typeshed stubs started to use this mechanism. However, I was assured at the time that their use would be limited to concrete types only — that no TypeVars would be used in these annotations. You're using a TypeVar here, which is why this isn't working.

I recommend that you use `__new__` overloads and eliminate the `__init__` overloads altogether. That way, you can stick with defined behaviors that all type checkers implement. It sounds like that's the direction you're heading already.
